core/account-id/src/lib.rs

```diff
@@ -0,0 +1,527 @@
+use std::{convert::TryFrom, str::FromStr};
+
+#[cfg(feature = "borsh")]
+use borsh::{BorshDeserialize, BorshSerialize};
+#[cfg(feature = "serde")]
+use serde::{self, de};
+
+pub const MIN_ACCOUNT_ID_LEN: usize = 2;
+pub const MAX_ACCOUNT_ID_LEN: usize = 64;
+
+#[derive(Debug, thiserror::Error)]
+pub enum ParseAccountError {
+    #[error("the value is too long for account ID")]
+    TooLong,
+    #[error("the value is too short for account ID")]
+    TooShort,
+    #[error("the value has invalid characters for account ID")]
+    Invalid,
+}
+
+/// Account identifier. Provides access to user's state.
+///
+/// *Note: Every owned Account ID has ensured its validity.*
+#[derive(
+    Eq,
+    Ord,
+    Hash,
+    Clone,
+    Debug,
+    PartialEq,
+    PartialOrd,
+    derive_more::Into,
+    derive_more::AsRef,
+    derive_more::Display,
+)]
+#[as_ref(forward)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "borsh", derive(BorshSerialize))]
+pub struct AccountId(
+    #[cfg_attr(feature = "serde", serde(deserialize_with = "serde_validate_account_id"))] Box<str>,
+);
+
+impl AccountId {
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn validate(account_id: impl AsRef<str>) -> Result<(), ParseAccountError> {
+        let account_id = account_id.as_ref();
+
+        if account_id.len() < MIN_ACCOUNT_ID_LEN {
+            Err(ParseAccountError::TooShort)
+        } else if account_id.len() > MAX_ACCOUNT_ID_LEN {
+            Err(ParseAccountError::TooLong)
+        } else {
+            // Adapted from https://github.com/near/near-sdk-rs/blob/fd7d4f82d0dfd15f824a1cf110e552e940ea9073/near-sdk/src/environment/env.rs#L819
+
+            // NOTE: We don't want to use Regex here, because it requires extra time to compile it.
+            // The valid account ID regex is /^(([a-z\d]+[-_])*[a-z\d]+\.)*([a-z\d]+[-_])*[a-z\d]+$/
+            // Instead the implementation is based on the previous character checks.
+
+            // We can safely assume that last char was a separator.
+            let mut last_char_is_separator = true;
+
+            for c in account_id.bytes() {
+                let current_char_is_separator = match c {
+                    b'a'..=b'z' | b'0'..=b'9' => false,
+                    b'-' | b'_' | b'.' => true,
+                    _ => return Err(ParseAccountError::Invalid),
+                };
+                if current_char_is_separator && last_char_is_separator {
+                    return Err(ParseAccountError::Invalid);
+                }
+                last_char_is_separator = current_char_is_separator;
+            }
+
+            (!last_char_is_separator).then(|| ()).ok_or(ParseAccountError::Invalid)
```

Comment:
last part should be at least 2 characters long I think, because it is root account that should be valid by itself

Reply:
I think it's worth noting that the previous behaviour didn't ensure this constraint. Rather, we check that the entire account ID is at least 2 characters *(this behaviour was retained)*.
https://github.com/near/nearcore/blob/f453895831ed423ec41628c034d5aae5e1164896/runtime/near-runtime-utils/src/lib.rs#L14-L15
And there's a chance that account IDs on, say testnet for instance, could currently violate this. While it's a check we should have, doing that here could block this PR.

Reply:
also I think all account ids that are 64 characters long have to be implicit (i. e. consist only of hex characters), see https://nomicon.io/DataStructures/Account.html